extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergTableAdapter.java

```diff
@@ -231,13 +251,29 @@ public TableDefinition definition(
     public TableDefinition definition(
             @Nullable final Snapshot tableSnapshot,
             @Nullable final IcebergInstructions instructions) {
-        final Snapshot snapshot = tableSnapshot != null ? tableSnapshot : table.currentSnapshot();
-        final Schema schema = snapshot != null ? table.schemas().get(snapshot.schemaId()) : table.schema();
+
+        final Snapshot snapshot;
+        final Schema schema;
+        final org.apache.iceberg.PartitionSpec partitionSpec;
+
+        if (tableSnapshot == null) {
+            synchronized (this) {
+                // Refresh only once and record the current snapshot, using its schema and spec.
+                refresh();
+                snapshot = table.currentSnapshot();
+                schema = snapshot != null ? schema(snapshot.schemaId()) : table.schema();
+                partitionSpec = table.spec();
```

Comment:
This needs some thought. Which schema is best, `table.schema()`, or the snapshot's one?
I *think* we want to do refresh, table schema, and table spec.